server.go

```diff
@@ -741,6 +741,7 @@ func (srv *Server) serveDNS(w *response) {
 	if w.udp != nil && cap(w.msg) == srv.UDPSize {
 		srv.udpPool.Put(w.msg[:srv.UDPSize])
 	}
+	tmpMsg := w.msg
```

Comment:
This is wrong, the `srv.udpPool.Put` above will cause data corruption. As a quick fix, move the entire
```go
if w.udp != nil && cap(w.msg) == srv.UDPSize {
	srv.udpPool.Put(w.msg[:srv.UDPSize])
}
w.msg = nil
```
block to immediately after the TSIG verification.

Reply:
I’d like something that doesn’t drop buffers at the returns above as a more premenent fix though.